### ISSUE TITLE
remove/delete method for zipfile objects

### ISSUE THREAD

**Issue by 4e9e3072-eb02-4932-b3a1-2e2df51207cb** [python/cpython#51067]
BPO | [6818](https://bugs.python.org/issue6818)
--- | :---
Nosy | @loewis, @rhettinger, @terryjreedy, @gustaebel, @tiran, @merwok, @sandrotosi, @serhiy-storchaka, @denfromufa, @yudilevi2
PRs | <li>python/cpython#19336</li><li>python/cpython#19358</li>
Files | <li>[delete.tar.gz](https://bugs.python.org/file15199/delete.tar.gz "Uploaded as application/x-gzip at 2009-10-26.07:15:09 by victorlee129"): Includes 2 files,the one named delete.py is the main file.</li><li>[zipfile.remove.patch](https://bugs.python.org/file21063/zipfile.remove.patch "Uploaded as text/plain at 2011-03-09.20:54:29 by ubershmekel"): bugs, docs and tests</li><li>[zipfile.remove.2.patch](https://bugs.python.org/file21188/zipfile.remove.2.patch "Uploaded as text/plain at 2011-03-15.01:08:08 by ubershmekel"): improved patch for zipfile.remove</li><li>[tricky.zip](https://bugs.python.org/file38231/tricky.zip "Uploaded as application/zip at 2015-02-25.10:57:18 by @serhiy-storchaka")</li><li>[mywork.patch](https://bugs.python.org/file39061/mywork.patch "Uploaded as text/plain at 2015-04-15.23:54:39 by Dave Sawyer")</li><li>[zipfile_filter.patch](https://bugs.python.org/file39063/zipfile_filter.patch "Uploaded as text/plain at 2015-04-16.02:15:06 by Dave Sawyer"): renamed, fixed comment, removed some debugging code</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2009-09-02.04:42:10.018>
labels = ['type-feature', 'library']
title = 'remove/delete method for zipfile/tarfile objects'
updated_at = <Date 2020-04-04.10:40:28.415>
user = 'https://bugs.python.org/rossmclendon'
```

bugs.python.org fields:
```python
activity = <Date 2020-04-04.10:40:28.415>
actor = 'yudilevi'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2009-09-02.04:42:10.018>
creator = 'rossmclendon'
dependencies = []
files = ['15199', '21063', '21188', '38231', '39061', '39063']
hgrepos = ['305']
issue_num = 6818
keywords = ['patch']
message_count = 36.0
messages = ['92154', '92155', '92156', '92158', '92164', '92172', '92177', '94475', '94478', '109158', '109160', '130299', '130388', '130463', '130938', '140680', '159418', '160533', '160534', '192574', '229801', '229893', '236565', '236568', '240336', '240345', '240394', '240422', '241181', '241192', '241201', '241346', '266374', '365755', '365756', '365757']
nosy_count = 17.0
nosy_names = ['loewis', 'rhettinger', 'terry.reedy', 'lars.gustaebel', 'christian.heimes', 'rossmclendon', 'eric.araujo', 'ubershmekel', 'victorlee129', 'sandro.tosi', 'chroipahtz', 'serhiy.storchaka', 'Arthur.Darcet', 'Dave Sawyer', 'gambl', 'denfromufa', 'yudilevi']
pr_nums = ['19336', '19358']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue6818'
versions = ['Python 3.5']
```

</p></details>

### Linked PRs
* gh-103033
* gh-134627

**Comment by buhtz** [comment]
I'm a bit overwhelmed about all the comments and linked Issues and PR's.
Does anyone has an overview about the current status of that Issue?

**Comment by merwok** [comment]
Read the four longest comments and ignore the small ones about process.

issues and possible approaches: https://github.com/python/cpython/issues/51067#issuecomment-1093479277
one approach: https://github.com/python/cpython/issues/51067#issuecomment-1093479282
another opnion: https://github.com/python/cpython/issues/51067#issuecomment-1093479287
another approach with a noted flaw: https://github.com/python/cpython/issues/51067#issuecomment-1093479293

The one PR that’s linked and open is from the author of the latest comment: https://github.com/python/cpython/pull/19358

So the status is to convince core devs that the approach is worth it, then ensure the PR contains complete docs and tests and applies to the current main branch, and get reviews.

**Comment by danny0838** [comment]
I have updated the code based on [yudilevi2](https://github.com/yudilevi2)'s version and added tests and docs. Should I discuss in that thread or open a new PR?

**Comment by merwok** [comment]
Discussions about the bug itself and the general idea of how to fix it should be here.
PR discussions can focus on whether the code is doing everything right to implement the solution agreed on 🙂

**Comment by danny0838** [comment]
Some folks are concerned about that, if a `remove` implementation handles file shrinking (which implies moving the data of internal file entries), and something bad happens during the moving, will lead to file corruption, as @serhiy-storchaka, @DaveSawyer, and @yudilevi2's comment have mentioned.

I will argue that **this will also happen in most in-place file modification scenarios**. A good similar example is appending a file to a ZIP, in which case the appended file data will first overwrite the central directory, and the ZIP will be corrupted if there is an interruption before the final new central directory data is written. In other words: if one isn't so concerned that appending a file may corrupt the ZIP, he shouldn't be so concerned about removing, either.

Although there is always a more sophisticated (and likely more indirect) technique to make an "in-place" file modification more secure, such as cloning data to an intermediate temporary file, I think it's reasonable enough for an official and basic implementation to work in the way.

**Comment by danny0838** [comment]
Pardon. Is there still something that prevents the patch from implemented. It seems that lacking tests and docs is the only thing for the previous one?

**Comment by merwok** [comment]
See the top post for PR link

**Comment by distantvale** [comment]
@merwok yup, sorry - so indeed what else is missing? I can work on that

**Comment by narugo1992** [comment]
This code not work for me on python3.8.1 ubuntu20.04

When you use mode `a` to remove the file inside, you'll find the `FFF.md` which should be delete is still in the `f.filelist`.

```python
import zipfile


# just copied from current commit's change
class ZFile(zipfile.ZipFile):
    def remove(self, zinfo_or_arcname):
        """Remove a member from the archive."""

        if self.mode not in ('w', 'x', 'a'):
            raise ValueError("remove() requires mode 'w', 'x', or 'a'")
        if not self.fp:
            raise ValueError(
                "Attempt to write to ZIP archive that was already closed")
        if self._writing:
            raise ValueError(
                "Can't write to ZIP archive while an open writing handle exists"
            )

        # Make sure we have an existing info object
        if isinstance(zinfo_or_arcname, zipfile.ZipInfo):
            zinfo = zinfo_or_arcname
            # make sure zinfo exists
            if zinfo not in self.filelist:
                raise KeyError(
                    'There is no item %r in the archive' % zinfo_or_arcname)
        else:
            # get the info object
            zinfo = self.getinfo(zinfo_or_arcname)

        return self._remove_members({zinfo})

    def _remove_members(self, members, *, remove_physical=True, chunk_size=2 ** 20):
        """Remove members in a zip file.
        All members (as zinfo) should exist in the zip; otherwise the zip file
        will erroneously end in an inconsistent state.
        """
        fp = self.fp
        entry_offset = 0
        member_seen = False

        # get a sorted filelist by header offset, in case the dir order
        # doesn't match the actual entry order
        filelist = sorted(self.filelist, key=lambda x: x.header_offset)
        for i in range(len(filelist)):
            info = filelist[i]
            is_member = info in members

            if not (member_seen or is_member):
                continue

            # get the total size of the entry
            try:
                offset = filelist[i + 1].header_offset
            except IndexError:
                offset = self.start_dir
            entry_size = offset - info.header_offset

            if is_member:
                member_seen = True
                entry_offset += entry_size

                # update caches
                self.filelist.remove(info)
                try:
                    del self.NameToInfo[info.filename]
                except KeyError:
                    pass
                continue

            # update the header and move entry data to the new position
            if remove_physical:
                old_header_offset = info.header_offset
                info.header_offset -= entry_offset
                read_size = 0
                while read_size < entry_size:
                    fp.seek(old_header_offset + read_size)
                    data = fp.read(min(entry_size - read_size, chunk_size))
                    fp.seek(info.header_offset + read_size)
                    fp.write(data)
                    fp.flush()
                    read_size += len(data)

        # Avoid missing entry if entries have a duplicated name.
        # Reverse the order as NameToInfo normally stores the last added one.
        for info in reversed(self.filelist):
            self.NameToInfo.setdefault(info.filename, info)

        # update state
        if remove_physical:
            self.start_dir -= entry_offset
        self._didModify = True

        # seek to the start of the central dir
        fp.seek(self.start_dir)


with ZFile('test_zip.zip', 'w') as f:
    f.write('README.md', '1/R.md')
    f.write('README.md', 'FFF.md')

with ZFile('test_zip.zip', 'a') as f:
    f.remove('FFF.md')

with ZFile('test_zip.zip', 'r') as f:
    print(f.getinfo('1/R.md'))
    print(f.getinfo('FFF.md'))

```

**Comment by danny0838** [comment]
@narugo1992

This is due to a bug in Python < 3.8.7 that zipfile.ZipFile is not truncated in 'a' mode when closed. You have to patch `ZipFile._write_end_record` to use in an affected Python version like this:

```Python
if _sys.version_info < (3, 8, 7):
    # Fix an issue causing zip not truncated
    class ZipFile(zipfile.ZipFile):
        def _write_end_record(self):
            super()._write_end_record()
            if self.mode == "a":
                self.fp.truncate()
            self.fp.flush()
    zipfile.ZipFile = ZipFile
```

**Comment by withnale** [comment]
This issue has been open long time. Is there any appetite for someone to get this working?

**Comment by danny0838** [comment]
Submitted a new PR #134627 for this issue.

**Comment by gpshead** [comment]
I raised a question on one of the `zipfile` PRs regarding `remove` & `repack` that was pointed out made more sense on the issue:

"""A higher level question: Why would we want to maintain advanced features that are not used by most people within the standard library at all?  This code existing in the stdlib creates a maintenance burden and is the slowest possible way to get features and their resulting trail of bugfixes to people who might want them. All features have an ongoing cost.

Is there a real _need_ for these zipfile features to not simply be advanced ones available in a PyPI package?  @jaraco FYI

I appreciate the enthusiasm for implementing something interesting. I'm not sure we actually want to maintain that within the CPython project though.  Are there compelling use cases for these features to be part of Python rather than external?""" - https://github.com/python/cpython/pull/134627#issuecomment-2924102128

Please don't read this as a blanket "No", I am posing it because it is something we should ask ourselves before adding maybe complicated features.  `zipfile` is, unfortunately, complicated given the messy nature of the file format so it is worth understanding if adding something like this is a maintenance burden or not, if bugs in the new feature impact other things or are isolated to the new areas, and if people are not better served by having this just be in a PyPI package, including if there are things we could do to the zipfile module APIs to enable this to be a more natural PyPI package add-on built on top of the stdlib.

**Comment by danny0838** [comment]
As an essential component of **CRUD**, I don't think there is not too much need to debate on whether `remove()` is needed.

Can you imagine that the `os` or `sqlite3` API that does not support data removal at all? That's why this issue has been raised here for over 15 years. Most things people are debating here is the how to elegantly implement it, not whether it should be implemented.

Another reason that this should be implemented in the stdlib is that it's highly incorporated with the module internal. For example, the implementation in PR #134627 accesses and/or changes `ZipFile._lock`, `ZipFile._writing`, `ZipFile.header_offset`, `ZipFile._end_offset`, etc. They are all hidden property and not guaranteed to be stable. If the feature is implemented as a standalone package depending on the `zipfile` package, it's at high risk of breaking in feature versions. OTOH implementing a fresh ZIP library from zero just for the `remove()` feature is likely an overkill.

**Comment by distantvale** [comment]
IMO the design isn't that complicated, and it's a feature that's been requested for years, some devs probably just gave up on it since it's been so long.
It's also a pretty basic feature, with a very well defined expected result. It will also probably be used more than some other obscure functions.
I've implemented my own version of it years ago, been merging it locally every few officialy releases, and it works like a charm for years.

**Comment by distantvale** [comment]
So? What's the verdict here? Better we push this forward before the PR becomes stale again.

**Comment by emmatyping** [comment]
> IMO the design isn't that complicated

I disagree with this. I think the API of `.remove` is deceptively simple. Sure it's just a single method. But I think the semantics of what that method actually *does* is where there is a lot of complexity. For example, last I checked the current PR will remove all space between existing entries listed in the central directory. To be clear, I do not highlight this as an issue with the current implementation. But it is a choice that needs to be made. The zip format allows arbitrary data between file entries, which as far as I am aware, zipfile has so far honored. I *expect* it would be fine to compact things in this case, but out of an abundance of caution I would like to see this tested in a third party package that sees usage outside of the stdlib before having it merged into CPython.

> So? What's the verdict here? 

I think @gpshead and I are both hesitant to merge something in if it hasn't seen usage because there are design choices that need to be used in the wild.

We also are waiting on @jaraco to give his opinion.

So based on all of the above, I disagree with:
> Better we push this forward before the PR becomes stale again.

**Comment by danny0838** [comment]
> > IMO the design isn't that complicated
> 
> I disagree with this. I think the API of `.remove` is deceptively simple. Sure it's just a single method. But I think the semantics of what that method actually _does_ is where there is a lot of complexity.

> For example, last I checked the current PR will remove all space between existing entries listed in the central directory.

This has been changed 2 weeks ago since c80d21bf991dff0369178f504212642f4000f769.

> To be clear, I do not highlight this as an issue with the current implementation. But it is a choice that needs to be made. The zip format allows arbitrary data between file entries, which as far as I am aware, zipfile has so far honored. I _expect_ it would be fine to compact things in this case, but out of an abundance of caution I would like to see this tested in a third party package that sees usage outside of the stdlib before having it merged into CPython.

This should better be said 2 weeks ago, so that we know if there's something to be determined, considered, or improved, instead of waiting for nothing for weeks.

> 
> > So? What's the verdict here?
> 
> I think [@gpshead](https://github.com/gpshead) and I are both hesitant to merge something in if it hasn't seen usage because there are design choices that need to be used in the wild.

So in brief you are not going to implement anything until it has been used widely as a 3rd party lib?

**Comment by jaraco** [comment]
It's been my experience that until there's something in stdlib, only a small subset of the users with the highest need will even discover or use a third-party lib, and thus most feedback will be delayed until there's a version in stdlib, especially for extensions of existing functionality like proposed here.

Regarding PyPI previews of features/functionality, I've found the model used by zipp and others to be particularly time consuming and challenging. It would be preferable to have one source of truth, possibly with backports (more like what configparser has). Keeping two versions of the functionality in sync can add significantly to the maintenance burden.

I do appreciate the API of the third-party `zipremove`, which provides a nice, consistent API for the consumer, as long as that wrapper doesn't conflict with another compatibility wrapper (such as [zipp uses internally](https://github.com/jaraco/zipp/blob/main/zipp/compat/overlay.py)). The risk of recommending a third-party shim for each (significant) feature is that there's a proliferation of such features that are mutually incompatible.

I do see there's quite a bit of code and logic to maintain (over 400 lines added to the zipfile module), but it also sounds as if there are significant interest from the community that any defects would find volunteers to aid in the maintenance.

My instinct is that this feature is closely related to the essence of a zipfile and would have real value in the stdlib. The proposed implementation is highly modular (doesn't interact with the existing functionality), so seems safe to add.

Especially given that we're early in the Python 3.15 cycle and that a third-party implementation exists for older Pythons, I'd like to see something merged or at least clarity on what could be accepted.

> I've implemented my own version of it years ago, been merging it locally every few officialy releases, and it works like a charm for years.

> So? What's the verdict here? Better we push this forward before the PR becomes stale again.

@distantvale I appreciate your passion, but do keep in mind that we're all volunteers here and the Python stdlib has a very high bar for acceptance of changes because of the wide audience it supports. Any change that's added is very sticky and more difficult than other systems to iterate. Given that you've been maintaining your own implementation of this feature, I'd appreciate your review on Danny's proposed implementation, and especially on the tradeoffs and open questions raised there (default for `strict_descriptor`, is `copy()` necessary) and any other perspective you have on the approach.

**Comment by cmaloney** [comment]
re: `.delete()` and `.repack()` how do they interact with the "reproducibility" of artifacts?

In particular, there is a lot of work that has been going on to make "reproducible builds" and my worry is implementers, to save compute/runtime, will start using "remove" (rather than rewriting a zip from scratch excluding a file), and that will mean their artifacts aren't reproducible as an add + remove doesn't produce a byte-for-byte identical zipfile to one where those operations never happened. 

I don't think adding and removing a file from a zipfile needs to produce the same bytes, and a lot of efficiency exists that they don't, but in the context of `.repack()` it seems like a case where giving a tool to ensure reproducibility would make sense and be useful. In particular, two zipfiles with an arbitrary string of add/remove oerations on them that in final state have the same list of files containing the same data should be byte-for-byte identical

**Comment by danny0838** [comment]
@cmaloney 

> re: `.delete()` and `.repack()` how do they interact with the "reproducibility" of artifacts?
> 
> In particular, there is a lot of work that has been going on to make "reproducible builds" and my worry is implementers, to save compute/runtime, will start using "remove" (rather than rewriting a zip from scratch excluding a file), and that will mean their artifacts aren't reproducible as an add + remove doesn't produce a byte-for-byte identical zipfile to one where those operations never happened.
> 
> I don't think adding and removing a file from a zipfile needs to produce the same bytes, and a lot of efficiency exists that they don't, but in the context of `.repack()` it seems like a case where giving a tool to ensure reproducibility would make sense and be useful. In particular, two zipfiles with an arbitrary string of add/remove oerations on them that in final state have the same list of files containing the same data should be byte-for-byte identical

`remove` and `repack` don't touch unrelated bytes. But it's expectable that file order and mtime would differ if handled with add/removing rather than regenerating the archive.

**Comment by merwok** [comment]
Personally I wouldn’t worry about packagers both using delete/repack and wanting reproducible builds. The latter is getting well documented and more widespread. It would be convoluted to imaging packagers using python zipfile and delete/repack instead of building then zipping with a well-know zip utility and appropriate options (umask, source date epoch, etc.)

**Comment by distantvale** [comment]
I agree that bytes unrelated to the direct operation should be left untouched - implicit changes are a pandora's box.

**Comment by distantvale** [comment]
@jaraco np - will review the PR.

**Comment by distantvale** [comment]
@danny0838 before I dive in into all the details, can you please elaborate a bit about the choice of making removal a 2 step process?It adds a bit of complexity, especially since it requires some heuristics during repack.
Also, what happens if the user calls remove() without calling repack() and then flushes/closes the file?

Btw lemme know if this discussion should move to the PR thread - not sure where the proper way to discuss these things is.

**Comment by danny0838** [comment]
@distantvale

In general I think it's more appropriate to discuss the implementation details about a specific PR in its own thread and the general idea about an issue in the issue thread. But whatever...

If the archive is closed when a `remove` is done without `repack`, then only the entry in the central directory is removed, while the local entry remains there. This should have been already documented clearly.

Regarding why `remove` and `repack`: it should have already been mentioned in https://github.com/python/cpython/pull/103033#issuecomment-1490705652 and provided in the top of the PR. Also the older implementation #103033 is still there, maybe you should ask why it has been kept undiscussed for years...

Some more my personal opinions:

1. **Less design consideration to deal with multiple entries**
 
   Although the older `remove` implementation supports removing + repacking at once, it handles only one entry and is especially imperformant when one wants to remove multiple entries, which is expected to be a common scenario.

   Even though it can be reworked to support removing multiple entries at once, like the hidden `_remove_members` does, there's still lots of design considerations to deal with to make it a public API. What the API should look like? Should it accept any iterable, a list, or maybe a filter function? Should all sub entries be removed together when removing a folder? And so on... This is probably one of the main reason why it has been left there for such a long time...

   Most current `ZipFile` APIs are rather low-level. For example writing does not guard from writing to an existing entry name, or writing content to a directory path. This is even more true if we also consider `copy` and `rename`: Should copying/renaming a file to a folder be forbidden? And so forth...

   OTOH, `remove` + `repack` (+ `copy`) are more elemental and consistent with the existing APIs that each handles one entry at a time. One can easily write a high level API in any way he likes based on them.

2. **More flexible by allowing delayed repacking**

   Keeping `remove` simple quick and safe and allow `repack` to strip stale entries without providing more details afterwards provides more flexibility. For example, one may want to quickly remove entries through a web app and let the repacking to happen at some system idle time (like Git house keeping does). While one can still provide the *removed* ZipInfos to `repack` if he doesn't want any heuristics.

   It also allows one to at least do the removing (and defer the repacking) when repacking is not available, such as when the underlying stream is unseekable, or when the zip file is opened in `'w'`/`'x'` mode (at least before the issue where they don't support truncating is fixed).

   Both are clear additions that the older design doesn't support.

**Comment by distantvale** [comment]
@danny0838 I'll move my feedback there to the PR thread

**Comment by jaraco** [comment]
It sounds to me like the discussion may have reach conclusion. Are there outstanding concerns before accepting #134627 ?

**Comment by danny0838** [comment]
I'm fine with the propose. However the review of @emmatyping is still pending and there's not yet a comment about the code, doc, or tests from @distantvale besides arguing whether the older API implementation should be used. I'd like to know if there's still an issue or something to be improved.

**Comment by emmatyping** [comment]
I can try to review the PR this week. I don't think my review should be a blocker though.

**Comment by danny0838** [comment]
It has been long for this PR. [The package version](https://github.com/danny0838/zipremove) has 13k downloads while no issue has been raised. Is there a chance for this to be merged in 3.15, or is there still any major issue preventing this from being merged?

**Comment by distantvale** [comment]
Is there anything missing from this PR in order to move forward?

**Comment by gpshead** [comment]
thanks for the contributions and discussion, danny0838's version of this has been merged.
